sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamIOPushDownRule.java

```diff
@@ -0,0 +1,236 @@
+        SelectHelpers.getOutputSchema(ioSourceRel.getBeamSqlTable().getSchema(), resolved);
+    RelDataType calcInputType =
+        CalciteUtils.toCalciteRowType(newSchema, ioSourceRel.getCluster().getTypeFactory());
+
+    BeamIOSourceRel newIoSourceRel = ioSourceRel.copy(calcInputType, new ArrayList<>(usedFields));
+    relBuilder.push(newIoSourceRel);
+
+    List<RexNode> newProjects = new ArrayList<>();
+    List<RexNode> newFilter = new ArrayList<>();
+    // Ex: let's say the original fields are (number before each element is the index):
+    // {0:unused1, 1:id, 2:name, 3:unused2},
+    // where only 'id' and 'name' are being used. Then the new calcInputType should be as follows:
+    // {0:id, 1:name}.
+    // A mapping list will contain 2 entries: {0:1, 1:2},
+    // showing how used field names map to the original fields.
+    List<Integer> mapping =
+        resolved.getFieldsAccessed().stream()
+            .map(FieldDescriptor::getFieldId)
+            .collect(Collectors.toList());
+
+    // Map filters to new RexInputRef
+    for (RexNode filter : projectFilter.right) {
+      newFilter.add(reMapRexNodeToNewInputs(filter, mapping));
+    }
+    // Map projects to new RexInputRef
+    for (RexNode project : projectFilter.left) {
+      newProjects.add(reMapRexNodeToNewInputs(project, mapping));
+    }
+
+    relBuilder.filter(newFilter);
+    relBuilder.project(newProjects, calc.getRowType().getFieldNames()); // Preserve named projects
+
+    RelNode result = relBuilder.build();
+
+    call.transformTo(result);
+  }
+
+  /**
+   * Given a {@code RexNode}, find all {@code RexInputRef}s a node or it's children nodes use.
+   *
+   * @param inputRowType {@code RelDataType} used for looking up names of {@code RexInputRef}.
+   * @param startNode A node to start at.
+   * @param usedFields Names of {@code RexInputRef}s are added to this list.
+   */
+  @VisibleForTesting
+  void findUtilizedInputRefs(RelDataType inputRowType, RexNode startNode, Set<String> usedFields) {
+    Queue<RexNode> prerequisites = new ArrayDeque<>();
+    prerequisites.add(startNode);
+
+    // Assuming there are no cyclic nodes, traverse dependency tree until all RexInputRefs are found
+    while (!prerequisites.isEmpty()) {
+      RexNode node = prerequisites.poll();
+
+      if (node instanceof RexCall) { // Composite expression, example: "=($t11, $t12)"
+        RexCall compositeNode = (RexCall) node;
+
+        // Expression from example above contains 2 operands: $t11, $t12
+        prerequisites.addAll(compositeNode.getOperands());
+      } else if (node instanceof RexInputRef) { // Input reference
+        // Find a field in an inputRowType for the input reference
+        int inputFieldIndex = ((RexInputRef) node).getIndex();
+        RelDataTypeField field = inputRowType.getFieldList().get(inputFieldIndex);
+
+        // If we have not seen it before - add it to the list (hash set)
+        usedFields.add(field.getName());
+      }
```

Comment:
Can you encounter another node type? If so, why doesn't it need to be handled, and can you assert that it is the expected type? If not, can you add an `else` that throws an exception?

Reply:
Here is a list of node types present in Calcite:
`RexCall`, `RexFieldAccess`, `RexLiteral`, `RexRangeRef`, `RexVariable`.

- [RexFieldAccess](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexFieldAccess.html) can only be found in queries like: 
```
example the expression emp.deptno in

SELECT ename
 FROM dept
 WHERE EXISTS (
     SELECT NULL
     FROM emp
     WHERE emp.deptno = dept.deptno
     AND gender = 'F')
```
which does not happen in `BeamIOSourceRel -> Calc`, but somewhere  in a `Join` as far as I can tell.
- [RexRangeRef](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexRangeRef.html) only used internally by calcite.
```
Regular rex trees do not contain this construct.
```
- [RexVariable](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexVariable.html) has following sub-classes:
  - [RexCorrelVariable](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexCorrelVariable.html) Correlating variables are introduced when performing nested loop joins, so should not be relevant at this stage.
  - [RexSlot](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexSlot.html) is extended by `RexInputRef`, and `RexLocalRef` (substituted by `RexProgram#split`).
  - [RexDynamicParam](https://calcite.apache.org/apidocs/org/apache/calcite/rex/RexDynamicParam.html) May need to be handled, but have not seen it appear in any existent tests.

Reply:
Modified to throw an exception when unhandled  node is encountered.